### PR TITLE
Prometheus: Add speculative block metrics

### LINKED ISSUE #1359: Add idle data to nodeos 5.0 prometheus exporter
In debugging https://github.com/AntelopeIO/leap/issues/1349, I asked Kevin if there are counters in the leap 5.0 prometheus exporter that capture the same content as these log lines:

```
Block #317926278 trx idle: 237us out of 199964us, success: 3, 578us, fail: 230, 197119us, transient: 0, 0us, other: 2030us
```

Seems no.  So this is a request to add counters.

### PR DESCRIPTION
**Pull request by heifner**
Added speculative block metrics:

total_time_us_produced_block - total time for produced block
idle_time_us_produced_block - idle time for produced block
num_success_trx_produced_block - number of successful transactions in produced block
success_trx_time_us_produced_block - time for successful transactions in produced block
num_failed_trx_produced_block - number of failed transactions during produced block
fail_trx_time_us_produced_block - time for failed transactions during produced block
num_transient_trx_produced_block - number of transient transactions during produced block
transient_trx_time_us_produced_block -time for transient transactions during produced block
other_time_us_produced_block - all other unaccounted time during produced block
block_num - current block number
blocks_speculative_num - number of speculative blocks created
total_time_us_speculative_block - total time for speculative block
idle_time_us_speculative_block - idle time for speculative block
num_success_trx_speculative_block - number of successful transactions in speculative block
success_trx_time_us_speculative_block - time for successful transactions in speculative block
num_failed_trx_speculative_block - number of failed transactions during speculative block
fail_trx_time_us_speculative_block - time for failed transactions during speculative block
num_transient_trx_speculative_block - number of transient transactions during speculative block
transient_trx_time_us_speculative_block - time for transient transactions during speculative block
other_time_us_speculative_block - all other unaccounted time during speculative block

Added the following to produced blocks:

nodeos_produced_elapsed_us_total - total produced blocks elapsed time
nodeos_produced_us_total - total produced blocks total time

Added the following to incoming blocks:

nodeos_incoming_elapsed_us_total - total incoming blocks elapsed time
nodeos_incoming_us_total - total incoming blocks total time
nodeos_incoming_us_block_latency - total incoming block latency

Resolves #1359

### CODE REVIEW

**Review by greg7mdp — APPROVED** [review]

File: plugins/prometheus_plugin/metrics.hpp [hunk 1]
```diff
@@ -91,11 +118,36 @@ struct catalog_type {
...
+       , block_other_time_us_speculative_block(build<Counter>("other_time_us_speculative_block", "all other unaccounted time during speculative block"))
        , trxs_incoming_total(build<Counter>("trxs_incoming_total", "number of incoming transactions"))
        , cpu_usage_us_incoming_block(cpu_usage_us.Add({{"block_type", "incoming"}}))
+       , total_elapsed_time_us_incoming_block(build<Counter>("nodeos_incoming_elapsed_us_total", "total incoming blocks elapsed time"))
+       , total_time_us_incoming_block(build<Counter>("nodeos_incoming_us_total", "total incoming blocks total time"))
        , net_usage_us_incoming_block(net_usage_us.Add({{"block_type", "incoming"}}))
+       , latency_us_incoming_block(build<Counter>("nodeos_incoming_us_block_latency", "total incoming block latency"))
        , blocks_incoming(build<Counter>("blocks_incoming", "number of incoming blocks"))
        , bytes_transferred(build<Counter>("exposer_transferred_bytes_total",
                                           "total number of bytes for responses to prometheus scape requests"))
```
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/prometheus_plugin/metrics.hpp]: Done

**Review by greg7mdp — APPROVED** [review]

File: plugins/prometheus_plugin/metrics.hpp [hunk 2]
```diff
@@ -118,25 +166,49 @@ struct catalog_type {
...
+      blk_metrics.block_other_time_us_block.Increment(metrics.block_other_time_us);
+   }
+
    void update(const producer_plugin::produced_block_metrics& metrics) {
       unapplied_transactions_total.Increment(metrics.unapplied_transactions_total);
       blacklisted_transactions_total.Increment(metrics.blacklisted_transactions_total);
       subjective_bill_account_size_total.Increment(metrics.subjective_bill_account_size_total);
       scheduled_trxs_total.Increment(metrics.scheduled_trxs_total);
       trxs_produced_total.Increment(metrics.trxs_produced_total);
-      blocks_produced.Increment(1);
```
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/prometheus_plugin/metrics.hpp:157]: It is now done in `update(produced_metrics, metrics);`, line 223.

File: plugins/prometheus_plugin/metrics.hpp [hunk 3]
```diff
@@ -52,20 +52,42 @@ struct catalog_type {
    Gauge& last_irreversible;
    Gauge& head_block_num;
 
+   struct block_metrics {
+      Counter& num_blocks;
+      Gauge&   block_num;
```
Comment (dimas1185) [inline plugins/prometheus_plugin/metrics.hpp:57]: I find `block_num` and `num_blocks` confusing naming. On lines `124` and `125` there is explanation. I suggest to name them accordingly to be self-descriptive. Something like that or maybe you have better ideas:
`num_blocks` -> `produced_blocks_num`
`block_num` -> `cur_block_num`

Comment (heifner) [reply plugins/prometheus_plugin/metrics.hpp]: Renamed

**Review by dimas1185 — APPROVED** [review]

**Pull request merged**